File: Week_02-哈希表-集合-树-图/2.两数相加.py
#
# @lc app=leetcode.cn id=2 lang=python
#
# [2] 两数相加
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        h1, h2 = l1, l2
        tmp = 0
        while h1 or h2: # 需要遍历最长的数
            val = h1.val + h2.val + tmp # 值=l1 + l2 +进位
            h1.val, tmp = val%10, val//10
            if not h1.next and h2.next: h1.next = ListNode(0) # 如果l1太短，添节点0
            if not h2.next and h1.next: h2.next = ListNode(0) # 如果l2太短，添节点0
            if not h1.next and not h2.next:  # 如果两个一样短，判断是否有进位，有就添节点
                if tmp: h1.next = ListNode(tmp)
                return l1 # 两个一样短的时候必然是返回的时候
            h1 = h1.next
            h2 = h2.next
        
# @lc code=end

File: Week_02-哈希表-集合-树-图/2.两数相加.cpp
/*
 * @lc app=leetcode.cn id=2 lang=cpp
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
#if 0
        int carry = 0;
        int sum = 0;
        ListNode* newhead = new ListNode(0, nullptr);//创建新链表头
        ListNode* nownode =newhead;  
        while(l1||l2||carry) {
            sum = (l1?l1->val:0) + (l2?l2->val:0) + carry;
            carry = sum/10;
            ListNode* node = new ListNode(sum%10, nullptr);//不停地new ListNode，比较消耗时间
            nownode->next = node;
            nownode = nownode->next;
            l1 = l1?l1->next:l1;
            l2 = l2?l2->next:l2;
        }
        return newhead->next;//返回链表头指向的下一个节点
#endif
#if 1
        int carry = 0;
        int sum = 0;
        ListNode newhead(0, nullptr);  //直接定义对象
        ListNode* currentNode = &newhead;
        while (l1 || l2 || carry) {
            sum = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
            carry = sum/10;
            ListNode* node = new ListNode(sum%10, nullptr);  //定义对象指针
            currentNode->next = node;
            currentNode = currentNode->next;
            l1 = l1 ? l1->next : l1;
            l2 = l2 ? l2->next : l2;
        }
        return newhead.next;

#endif

#if 0
//国外高票代码，11行实现相同方法
        ListNode newhead(0, nullptr), *currentnode = &newhead;
        int carry = 0;
        while (l1 || l2 || carry) {
            std::cout<<carry<<std::endl;
            int sum = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
            carry = sum / 10;
            currentnode->next = new ListNode(sum % 10);
            currentnode = currentnode->next;
            l1 = l1 ? l1->next : l1;
            l2 = l2 ? l2->next : l2;
        }
        return newhead.next;
#endif         
    }


};
// @lc code=end

